packages/cardpay-sdk/sdk/exchange-rate.ts

```diff
@@ -11,6 +11,17 @@ const ten = new BN('10');
 export default class ExchangeRate {
   constructor(private layer2Web3: Web3) {}
 
+  async getUSDConverter(token: string): Promise<(amountInWei: string) => number> {
+    const oracle = await this.getOracleContract(token);
+    const usdRawRate = new BN((await oracle.methods.usdPrice().call()).price);
+    const oracleDecimals = Number(await oracle.methods.decimals().call());
+
+    return function convertToUsd(amountInWei: string): number {
```

Comment:
Wondering if this should be something like an instance of a class that can internally keep track of when this exchange rate was fetched.

Reply:
I like the approach. although naming it is probably unnecessary, since the name is lost. probably more concise to say:
```ts
  return (amountInWei: string): number => {
    ...
  }
```

Reply:
knowing when to check on chain for the latest rate is a bit tricky. You really want to stay in sync with the contract, otherwise you are going to get weird results if you get out of sync with what is happening on chain. Perhaps there an be an internal polling mechanism that queries the contract every 60 seconds. if it sees that the price is different it can update this function via closure (in which case the `=>` will become ver important).

Reply:
albeit you also want to be cause in that you only ever have one poller at a time (using this 2 should not result in 2 pollers--rather one poller that can update all the instances where it is being used)

Reply:
Also we should definitely bounce this idea above off @lukemelia to see if he has any insight around good approaches here.

Reply:
Completely different approach could be using ethereum events to know when the rate changes. The quirk here is that the addresses that you'll need to listen to the events will originate from contracts we don't control (but the SDK can hide the address details). When the rates change ethereum events are emitted (the annoyance is that it is two totally different events because we have to different exchange rate providers). I think, though, that you'd still need some kind of polling mechanism here, as luke was encountering issues with the `subscribe()` API from web3. In this scenario the timing would be very precise, as you'll have ethereum events that are emitted at exactly when the rate changes.